compiler/src/lnntoamm/Types.ts

```diff
@@ -0,0 +1,290 @@
+import { LPNode } from "../lp";
+import Scope from "./Scope";
+import { genName, TODO } from "./util";
+
+type Fields = {[name: string]: Type | null};
+export type GenericArgs = {[name: string]: Type | null};
+type TypeName = [string, TypeName[]];
+
+const parseFulltypename = (node: LPNode): TypeName => {
+  const name = node.get('typename').t.trim();
+  let genericTys: TypeName[] = [];
+  if (node.get('opttypegenerics').has()) {
+    const generics = node.get('opttypegenerics').get();
+    genericTys.push(parseFulltypename(generics.get('fulltypename')));
+    generics.get('cdr').getAll().map(n => n.get('fulltypename'));
+  }
+  return [name, genericTys];
+};
+
+export default abstract class Type {
+  name: string
+  alias: Type | null
+
+  constructor(
+    name: string,
+    alias: Type | null,
+  ) {
+    this.name = name;
+    this.alias = alias;
+  }
+
+  static fromInterfacesAst(ast: LPNode, scope: Scope): Type {
+    return Interface.fromAst(ast, scope);
+  }
+
+  static fromTypesAst(ast: LPNode, scope: Scope): Type {
+    return Struct.fromAst(ast, scope);
+  }
+
+  static generate(): Type {
+    return new Generated();
+  }
```

Comment:
Is this supposed to be replaced by the other classes? Can it be declared `abstract` too? Or should we just leave a comment?

Reply:
Not sure what you mean? This is just a generated type - a type that hasn't been declared or reduced to a concrete type yet.

Reply:
Got it. I didn't understand what this is at first. Calling "generate" on a type sounds like it should do something, not create a type constraining object.

Reply:
the object isn't intended to constrain another type - instead, it's simply a way of converting
```alan
let x = 'hello, world!';
```
into
```alan
let x: T = 'hello, world!';
```
which will allow `T` to get reduced into some type `U` once additional constraints are added. In this case, `T` is constrained to a `string` type, so internally it will delegate to the `string` Type. I couldn't think of a better way to describe this than "generate a type", so that's why I went with `generate`